js/blocks/components/fields.js

```diff
@@ -59,33 +60,31 @@ const Fields = ( { fields, parentBlockProps, parentBlock, rowIndex } ) => {
 		 * the block's setAttributes property, so that the block can save the value.
 		 * This function is passed to the control so that the control can save the value,
 		 * depending on whether the control is in a repeater row or not.
+		 * This gets the latest parentAttributes by using select( 'core/block-editor' ),
+		 * as the parentBlockProps.attributes in the outer scope may be outdated.
 		 *
 		 * @param {*} newValue The new control value.
 		 */
 		const onChange = ( newValue ) => {
-			const attr = { ...parentBlockProps.attributes };
-			const attribute = attr[ field.parent ];
-			const { setAttributes } = parentBlockProps;
-			const defaultRows = [ {} ];
+			const { clientId, setAttributes } = parentBlockProps;
+			const parentAttributes = select( 'core/block-editor' ).getBlockAttributes( clientId );
+			const attr = { ...parentAttributes };
 
 			if ( undefined === rowIndex ) {
 				// This is not in a repeater row.
 				attr[ field.name ] = newValue;
 				setAttributes( attr );
 			} else {
 				// This is in a repeater row.
+				const attribute = attr[ field.parent ];
+				const defaultRows = [ {} ];
 				const rows = ( attribute && attribute.rows ) ? attribute.rows : defaultRows;
 
-				/*
-				 * Copy the rows array, so the change is recognized.
-				 * @see https://github.com/WordPress/gutenberg/issues/7016#issuecomment-396094836
-				 */
-				const rowsCopy = rows.slice();
```

Comment:
It looks like this `rows.slice()` hack isn't needed anymore to allow saving repeater values. But more testing would be good.

Reply:
I tested saving all of the controls in a repeater, and didn't see an issue.

Reply:
I also tested, everything seemed to work.